internal/cache/cache.go

```diff
@@ -0,0 +1,111 @@
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+//go:generate mockgen -source=cache.go -package=cache -destination=mock_cache.go
+
+type Cache interface {
+	DeleteExpired()
+	Get(key string) (interface{}, bool)
+	Set(key string, value interface{})
+	StopCleanup()
+}
+
+type Item struct {
+	Object interface{}
+
+	Expiration int64
+}
+
+type cache struct {
+	mu         sync.RWMutex
+	items      map[string]Item
+	expiration time.Duration
+	stop       chan struct{}
+	wg         sync.WaitGroup
+}
+
+// New returns a new cache with the requested item expiration duration, which automatically
+// removes expired items at the cleanup interval.
+func New(expiration, cleanupInterval time.Duration) *cache {
+	c := &cache{
+		items:      make(map[string]Item),
+		expiration: expiration,
+		stop:       make(chan struct{}),
+	}
+
+	c.wg.Add(1)
+	go func(interval time.Duration) {
+		defer c.wg.Done()
+		c.cleanup(interval)
+	}(cleanupInterval)
+
+	return c
+}
+
+func (c *cache) cleanup(interval time.Duration) {
+	t := time.NewTicker(interval)
+
+	for {
+		select {
+		case <-t.C:
+			c.DeleteExpired()
+		case <-c.stop:
+			t.Stop()
+			return
+		}
+	}
+}
+
+// StopCleanup signals the cleanup goroutine to stop its ticker and return.
+func (c *cache) StopCleanup() {
+	close(c.stop)
+	c.wg.Wait()
+}
+
+// Set adds an item to the cache, replacing any existing item.
+func (c *cache) Set(key string, value interface{}) {
+	expiration := time.Now().Add(c.expiration).UnixNano()
+
+	c.mu.Lock()
+	c.items[key] = Item{
+		Object:     value,
+		Expiration: expiration,
+	}
+	c.mu.Unlock()
+}
+
+// Get fetches an item from the cache. It returns the item or nil and a bool indicating
+// whether the key was found.
+func (c *cache) Get(k string) (interface{}, bool) {
+	c.mu.RLock()
+	item, found := c.items[k]
```

Comment:
you can probably use defer here, instead of calling c.mu.RUnlock, it will also be safer

Reply:
I considered using `defer`, but I thought I'd keep this bit faster (but maybe I should have measured `defer`'s performance before doing such an optimisation). I'll revert this to using a `defer`, since it's safer and more readable.